zkevm-circuits/src/rlp_decoder_tables.rs

```diff
@@ -0,0 +1,764 @@
+    }
+
+    /// build from existed columns
+    pub fn build_from_columns(columns: &[Column<Fixed>]) -> Self {
+        assert!(columns.len() > 2);
+        Self {
+            table_tag: columns[0],
+            value: columns[1],
+            len_decodable: columns[2],
+            val_decodable: columns[3],
+        }
+    }
+
+    /// Get the row num of the table
+    pub fn table_size() -> usize {
+        256
+    }
+
+    /// Assign the values of the table to the circuit
+    pub fn load<F: Field>(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        challenges: &Challenges<Value<F>>,
+    ) -> Result<(), Error> {
+        let mut randomness = F::ZERO;
+        challenges.keccak_input().map(|r| randomness = r);
+
+        layouter.assign_region(
+            || format!("load rlp heading bytes"),
+            |mut region| {
+                (0..256).try_for_each(|i| {
+                    region.assign_fixed(
+                        || "table tag",
+                        self.table_tag,
+                        i,
+                        || {
+                            Value::known(F::from(
+                                RlpDecoderFixedTableTag::HeadingByteValidTable as u64,
+                            ))
+                        },
+                    )?;
+                    region.assign_fixed(
+                        || "value",
+                        self.value,
+                        i,
+                        || Value::known(F::from(i as u64)),
+                    )?;
+                    region.assign_fixed(
+                        || "len_of_len valid",
+                        self.len_decodable,
+                        i,
+                        || {
+                            if i <= 55 {
+                                Value::known(F::ZERO)
+                            } else {
+                                Value::known(F::ONE)
+                            }
+                        },
+                    )?;
+                    region.assign_fixed(
+                        || "value valid",
+                        self.val_decodable,
+                        i,
+                        || {
+                            if i > 0x0 {
+                                Value::known(F::ONE)
+                            } else {
+                                Value::known(F::ZERO)
+                            }
+                        },
+                    )?;
+                    Ok(())
+                })
+            },
+        )?;
+        Ok(())
+    }
+}
+
+/// tag for shared fixed table
+#[derive(Clone, Copy, Debug)]
+pub enum RlpDecoderFixedTableTag {
+    /// All zero lookup data
+    Disabled,
+    /// Power of randomness: [1, r], [2, r^2],...
+    RMult,
+    /// 0 - 255
+    Range256,
+    /// Decode rule table for rlp tx
+    RlpDecoderTable,
+    /// Tx field switch table
+    TxFieldSwitchTable,
+    /// valid len/value heading byte
+    HeadingByteValidTable,
```

Comment:
Question for my understanding :) 

I see these 3 tables belong to different regions. Is it necessary to distinguish them with tags? 

I genuinely don't know!

Reply:
I think lookup uses columns which contains all regions it covers. like this:
<img width="417" alt="image" src="https://github.com/taikoxyz/zkevm-circuits/assets/99081233/7594e911-e9b8-4eb6-a905-d0e46691d146">

So, based on my understanding, I can setup region A/B/C separately, and lookup will take them all into account.
correct me if I am wrong @Brechtpd .